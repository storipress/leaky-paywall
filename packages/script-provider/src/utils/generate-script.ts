import { Effect, pipe } from 'effect'
import { oneLineTrim } from 'proper-tags'
import type { Context } from 'hono'
import esbuild from 'esbuild-wasm'
import { GraphqlService } from '../services/GraphqlService'
import { initEsbuild } from './esbuild-init'
import { getPaywallConfig } from './get-paywall-config'
import { CONFIG_VAR_NAME, PRODUCTION_DEBUG_URL, PRODUCTION_URL } from './constants'

const javascript = oneLineTrim
export function generateScript(c: Context, clientId: string) {
  return pipe(
    initEsbuild,
    Effect.andThen(getPaywallConfig(clientId)),
    Effect.flatMap((configValues) => {
      const config = JSON.stringify(configValues)
      const code = javascript`
            window.${CONFIG_VAR_NAME} = ${config};
            function insertScript(u) {
              let s=document.createElement('script');
              s.type='module';
              s.src=u;
              document.head.append(s);
            }
            insertScript('${PRODUCTION_URL}');
            if (window.location.search.includes('sp_debug=true')) {
              insertScript('${PRODUCTION_DEBUG_URL}');
            }
          `
      return Effect.promise(() => esbuild.transform(code, { loader: 'js', minify: true }))
    }),
    Effect.map((minified) => c.text(minified.code, 200, { 'content-type': 'text/javascript' })),
    Effect.provide(GraphqlService.layer(clientId)),
    Effect.catchTags({
      NotFoundError: () => Effect.succeed(c.text('Not Found', 404)),
      GraphqlError: (error) => {
        return pipe(Effect.logError(error), Effect.as(c.text('Internal Server Error', 500)))
      },
    }),
  )
}
