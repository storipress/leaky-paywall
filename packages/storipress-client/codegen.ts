import type { CodegenConfig } from '@graphql-codegen/cli'

const config: CodegenConfig = {
  schema: 'https://api.storipress.dev/graphql',
  documents: ['./queries/*.gql.ts'],
  ignoreNoDocuments: true, // for better experience with the watcher
  hooks: {
    afterAllFileWrite: ['prettier --write'],
  },
  generates: {
    './gql/': {
      preset: 'client',
      config: {
        useTypeImports: true,
        scalars: {
          JSON: 'string',
          EmailString: 'string',
          DateTime: 'string',
          Date: 'string',
        },
      },
    },
  },
}

export default config
