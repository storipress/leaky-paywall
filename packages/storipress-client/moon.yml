tasks:
  generate:
    command: gql-gen
    platform: node
    inputs:
      - queries/*.gql.ts
      - codegen.ts
    outputs:
      - gql/**
  generate-watch:
    extends: generate
    args: --watch
    local: true
