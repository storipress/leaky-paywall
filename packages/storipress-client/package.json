{
  "name": "storipress-client",
  "packageManager": "yarn@4.1.1",
  "devDependencies": {
    "@graphql-codegen/cli": "5.0.2",
    "@graphql-codegen/client-preset": "4.2.5",
    "@parcel/watcher": "2.4.1",
    "graphql": "^16.8.1"
  }
}
