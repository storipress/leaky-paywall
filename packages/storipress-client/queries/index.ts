export { SignInPaywall } from './SignIn.gql'
export { TrackSubscriberActivity } from './TrackSubscriberActivity.gql'
