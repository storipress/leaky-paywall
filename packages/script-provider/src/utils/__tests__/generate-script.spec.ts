import { HttpResponse, graphql, http } from 'msw'
import { setupServer } from 'msw/node'
import { SiteSubscriptionInfo } from 'storipress-client'
import { afterAll, beforeAll, expect, it, vi } from 'vitest'
import { Hono } from 'hono'
import { Effect, Logger, pipe } from 'effect'
import { generateScript } from '../generate-script'

vi.mock('esbuild-wasm', () => ({
  default: {
    transform: (x: unknown) => Promise.resolve(x),
  },
}))

vi.mock('../esbuild-init', async () => {
  const effect = await vi.importActual<{ Effect: typeof Effect }>('effect')
  return {
    initEsbuild: effect.Effect.void,
  }
})

const server = setupServer(
  http.get('https://api.storipress.dev/client/client_id/rest/v1/publication/state', () => {
    return HttpResponse.json({ state: 'online' })
  }),
  graphql.query(SiteSubscriptionInfo, () => {
    return HttpResponse.json({
      data: {
        siteSubscriptionInfo: {
          name: 'Test',
          description: 'Test description',
          logo: null,
          paywall_config: JSON.stringify({
            flags: {
              paywall: true,
              tracking: true,
            },
            freeLimit: {
              interval: 7,
              quota: 1,
            },
            pathPattern: null,
            all: false,
            clientId: '',
            logo: '',
            title: '',
            description: '',
            primaryColor: 'rgb(29 78 216)',
            dismissible: false,
          }),
        },
      },
    })
  }),
)

beforeAll(() => {
  server.listen()
})

afterAll(() => server.close())

it('can generate script', async () => {
  const app = new Hono().get('/', async (c) => {
    return pipe(
      generateScript(c, 'client_id'),
      Effect.provide(Logger.replace(Logger.defaultLogger, Logger.none)),
      Effect.runPromise,
    )
  })

  const res = await app.request('/')

  expect(res.status).toBe(200)
  await expect(res.text()).resolves.toMatchSnapshot()
})
