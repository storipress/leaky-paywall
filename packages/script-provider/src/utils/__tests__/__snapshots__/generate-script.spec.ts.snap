// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`can generate script 1`] = `"window.SP_PAYWALL = {"flags":{"paywall":true,"tracking":true},"freeLimit":{"interval":7,"quota":1},"paywallTrigger":{"type":"viewport","value":0.45},"pathPattern":null,"all":false,"clientId":"client_id","logo":"","title":"Test","description":"","primaryColor":"rgb(29 78 216)","dismissible":false};function insertScript(u) {let s=document.createElement('script');s.type='module';s.src=u;document.head.append(s);}insertScript('https://assets.stori.press/storipress/leaky-paywall.min.js?v=0.0.0-test');if (window.location.search.includes('sp_debug=true')) {insertScript('https://assets.stori.press/storipress/leaky-paywall-debug.min.js?v=0.0.0-test');}"`;
